lib/auth/accountrecovery.go

```diff
@@ -307,79 +271,28 @@ func (a *Server) verifyAuthnWithRecoveryLock(ctx context.Context, startToken typ
 	case trace.IsConnectionProblem(verifyAuthnErr):
 		log.Error(trace.DebugReport(verifyAuthnErr))
 		return trace.AccessDenied(verifyRecoveryBadAuthnErrMsg)
-
 	case verifyAuthnErr == nil:
-		// Reset attempt counter.
-		if err := a.DeleteUserRecoveryAttempts(ctx, startToken.GetUser()); err != nil {
-			log.Error(trace.DebugReport(err))
-		}
-
 		return nil
 	}
 
 	log.Error(trace.DebugReport(verifyAuthnErr))
-
-	lockedUntil, maxedAttempts, err := a.recordFailedRecoveryAttempt(ctx, startToken.GetUser())
-	switch {
-	case err != nil:
-		log.Error(trace.DebugReport(err))
-		return trace.AccessDenied(verifyRecoveryBadAuthnErrMsg)
-	case !maxedAttempts:
-		return trace.AccessDenied(verifyRecoveryBadAuthnErrMsg)
-	}
-
-	// Delete all tokens related to this user, to force user to restart the recovery flow.
-	if err := a.deleteUserTokens(ctx, startToken.GetUser()); err != nil {
```

Comment:
Should we keep this part? Does it have to do with recovery attempts too?

Reply:
Thank you for mentioning this.  I would like to hear @kimlisa's perspective on this removal.

This condition made no sense to me, having a reset failure should not allow any impacts to the account itself.  These tokens are used to generate new recovery codes, so potential impact cases are unlikely, but the behavior of allowing a potentially unauthorized actor to impact the account seems wrong to me.

I also made a recent relevant change to ensure these tokens are always deleted after use: https://github.com/gravitational/teleport/pull/36527

So in theory these tokens should be very short lived (unless the client was interrupted or otherwise stopped the token generation process for some reason)

I believe because of that, this was in error, but maybe @kimlisa is aware of something I am missing.

Reply:
The tokens being deleted in this line are tokens used to generate new recovery codes, not the recovery code itself.

> the particular token in use here will always be deleted

The recovery code is removed after use here: https://github.com/gravitational/teleport/blob/f356242a2c30ca28f6326d175adeb8ac22655706/lib/auth/accountrecovery.go#L146-L151

The tokens used for generating recovery codes will be deleted after successful use (as of the linked PR above), but I am not aware of a deletion process if they go unused (maybe there is a TTL I have not seen)

> but we won't delete any other recovery tokens to avoid a situation where a malicious user keeps making someone else recovery attempts fail through this side-effect?

This is correct as of this change.  But to ensure the potential impact is highlighted, it would only impact a user attempting to regenerate their recovery codes while an attacker was doing an invalid recovery attempt.  It is a very small risk, but I still believe this change is more technically correct.

@kimlisa Let me know if I have any misunderstandings, thank you!

Reply:
i didn't really understand what error the prior code had 😅 (i would very much like to understand though).

the start token used here expired in 3 hrs (it's the token that's part of the recovery URL we send to users email after they provided us with a valid recovery code)

in the prior code, the `delete token` block was only ran if the user reached max attempts at providing us their password/2ndfactor. i thought it appropriate then to [delete all user tokens](https://github.com/gravitational/teleport/blob/5f9803917c3daecb9caec434612be0e4b5f31231/lib/auth/usertoken.go#L44) so they can't `attempt anymore with this token` and the user is going to get locked anyways.

if the user tried to access the URL with the deleted token, it'll show something like "expired or invalid token", prompting the user to try another recovery code. 

i think if we didn't delete it, the way it was coded, it will allow the user to keep attempting?

but since we are removing max attempts and locks we no longer need the `delete tokens`, we just allow the user to keep trying it looks like?

Reply:
Thank you @kimlisa, I did not realize this was also deleting the token used in the email flow.  That said I believe this should still be removed, consider the following:

1. A victim is under targeted attack, their account is locked and they can not sign in.  They initiate a recovery process to unlock their account and attempt to gain immediate access.
2. The attacker is persistent and is simultaneously and continually doing false recovery attempts.  Resulting in the deletion of the initiated recovery email token.
3. The victim has no ability to recover their account unless they can get lucky enough to complete the entire process before the attacker can invalidate new tokens (or they end up using all their recovery codes trying).

Any token invalidation from an unauthenticated and untrusted party has some potential to lead to a denial of service.  For that reason this seems potentially risky with or without the recovery lockout removal.

> i think if we didn't delete it, the way it was coded, it will allow the user to keep attempting?

Maybe, this is not clear to me.  I expected to see this action as part of https://github.com/gravitational/teleport/blob/master/lib/auth/accountrecovery.go#L386, however I am not sure where this token is being deleted after use?  I just verified it's being deleted somewhere, and I don't see any way this line was the delete responsible to prevent reuse of the email token.

Let me know your thoughts and questions, thank you very much @kimlisa!

Reply:
got it, thanks for explaining!

> however I am not sure where this token is being deleted after use

if you've successfully authenticated with the start token, then we delete it afterwards [here](https://github.com/gravitational/teleport/blob/a15876724067837ba0449f06f01b9b0a25c370d0/lib/auth/accountrecovery.go#L283). i think it makes sense to delete it right afterwards in case user for some reason doesn't make it to completing the recovery (since it's a longer living token)

if you failed, you will get unlimited attempts (just confirming that's how it's currently going to work now)

We also delete all user tokens when we finish [creating new recovery codes](https://github.com/gravitational/teleport/blob/a15876724067837ba0449f06f01b9b0a25c370d0/lib/auth/accountrecovery.go#L491) (which is after `completing account recovery` (this portion just removes locks and allows you to reset password or second factor))

Reply:
Thank you, that makes sense!